tower-http/src/services/fs/serve_dir/mod.rs

```diff
@@ -350,6 +401,57 @@ where
     }
 }
 
+impl<ReqBody, F, FResBody> Service<Request<ReqBody>> for ServeDir<F>
+where
+    F: Service<Request<ReqBody>, Response = Response<FResBody>, Error = Infallible> + Clone,
+    F::Future: Send + 'static,
+    FResBody: http_body::Body<Data = Bytes> + Send + 'static,
+    FResBody::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    type Response = Response<ResponseBody>;
+    type Error = Infallible;
+    type Future = InfallibleResponseFuture<ReqBody, F>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if let Some(fallback) = &mut self.fallback {
+            fallback.poll_ready(cx)
+        } else {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let future = self.try_call(req).map(|result| -> Result<_, Infallible> {
+            match result {
+                Ok(response) => Ok(response),
+                Err(err) => {
+                    tracing::error!(error = %err, "Failed to read file");
+
+                    let body =
+                        ResponseBody::new(Empty::new().map_err(|err| match err {}).boxed_unsync());
+                    let response = Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(body)
+                        .unwrap();
+                    Ok(response)
+                }
+            }
+        } as _);
```

Comment:
What does this cast do?

Reply:
It makes rust infer the right type for the closure. If you don't have it you get this error

```
error[E0308]: mismatched types
   --> tower-http/src/services/fs/serve_dir/mod.rs:439:39
    |
425 |         let future = self.try_call(req).map(|result: Result<_, _>| -> Result<_, Infallible> {
    |                                             ----------------------------------------------- the found closure
...
439 |         InfallibleResponseFuture::new(future)
    |         ----------------------------- ^^^^^^ expected fn pointer, found closure
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected struct `futures_util::future::Map<_, fn(Result<http::Response<serve_dir::ResponseBody>, std::io::Error>) -> Result<http::Response<serve_dir::ResponseBody>, Infallible>>`
               found struct `futures_util::future::Map<_, [closure@tower-http/src/services/fs/serve_dir/mod.rs:425:45: 425:92]>
```